cpp/src/row_conversion/row_conversion.cu

```diff
@@ -0,0 +1,1321 @@
+/*
+ * Copyright (c) 2020-2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <iostream>
+#include <iterator>
+#include <limits>
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/sequence.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <cudf/row_conversion.hpp>
+#include "cudf/types.hpp"
+#include "rmm/device_buffer.hpp"
+#include "thrust/iterator/counting_iterator.h"
+#include "thrust/iterator/transform_iterator.h"
+
+namespace cudf {
+
+namespace detail {
+
+static inline __host__ __device__ int32_t align_offset(int32_t offset, std::size_t alignment)
+{
+  return (offset + alignment - 1) & ~(alignment - 1);
+}
+
+/**
+ * Copy a simple vector to device memory asynchronously. Be sure to read
+ * the data on the same stream as is used to copy it.
+ */
+template <typename T>
+std::unique_ptr<rmm::device_uvector<T>> copy_to_dev_async(const std::vector<T> &input,
+                                                          rmm::cuda_stream_view stream,
+                                                          rmm::mr::device_memory_resource *mr)
+{
+  std::unique_ptr<rmm::device_uvector<T>> ret(new rmm::device_uvector<T>(input.size(), stream, mr));
+  CUDA_TRY(cudaMemcpyAsync(
+    ret->data(), input.data(), sizeof(T) * input.size(), cudaMemcpyHostToDevice, stream.value()));
+  return ret;
+}
+
+template <typename T>
+rmm::device_uvector<T> copy_to_dev_async2(const std::vector<T> &input,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::mr::device_memory_resource *mr)
+{
+  rmm::device_uvector<T> ret(input.size(), stream, mr);
+  CUDA_TRY(cudaMemcpyAsync(
+    ret.data(), input.data(), sizeof(T) * input.size(), cudaMemcpyHostToDevice, stream.value()));
+  return ret;
+}
```

Comment:
I believe there are already factories that do this.

Reply:
I'll look around. Do you know the name? This does seem a very common thing.

Reply:
[`make_device_uvector_async`](https://github.com/rapidsai/cudf/blob/a0b0eab8e0730a86a638d86aeef11088f12c935c/cpp/include/cudf/detail/utilities/vector_factories.hpp#L121-L131)
[cudf/detail/utilities/vector_factories.hpp](https://github.com/rapidsai/cudf/blob/branch-21.08/cpp/include/cudf/detail/utilities/vector_factories.hpp)